generative/maisi/scripts/diff_model_train.py

```diff
@@ -0,0 +1,292 @@
+            monai.transforms.Lambdad(
+                keys="top_region_index",
+                func=lambda x: torch.FloatTensor(json.load(open(x))["top_region_index"]),
+            ),
+            monai.transforms.Lambdad(
+                keys="bottom_region_index",
+                func=lambda x: torch.FloatTensor(json.load(open(x))["bottom_region_index"]),
+            ),
+            monai.transforms.Lambdad(
+                keys="spacing",
+                func=lambda x: torch.FloatTensor(json.load(open(x))["spacing"]),
+            ),
+            monai.transforms.Lambdad(keys="top_region_index", func=lambda x: x * 1e2),
+            monai.transforms.Lambdad(keys="bottom_region_index", func=lambda x: x * 1e2),
+            monai.transforms.Lambdad(keys="spacing", func=lambda x: x * 1e2),
+        ]
+    )
+
+    cache_rate = args.diffusion_unet_train["cache_rate"]
+    print(f"cache_rate: {cache_rate}")
+    train_ds = monai.data.CacheDataset(
+        data=train_files,
+        transform=train_transforms,
+        cache_rate=cache_rate,
+        num_workers=2,
+    )
+
+    num_images_per_batch = args.diffusion_unet_train["batch_size"]
+    print(f"num_images_per_batch -> {num_images_per_batch}.")
+    train_loader = ThreadDataLoader(train_ds, num_workers=6, batch_size=num_images_per_batch, shuffle=True)
+
+    unet = define_instance(args, "diffusion_unet_def").to(device)
+    unet = torch.nn.SyncBatchNorm.convert_sync_batchnorm(unet)
+
+    # Multi-GPU support
+    if torch.cuda.device_count() > 1:
+        unet = DistributedDataParallel(unet, device_ids=[device], find_unused_parameters=True)
+
+    if existing_ckpt_filepath == None:
+        print("training from scratch.")
+    else:
+        checkpoint_unet = torch.load(f"{existing_ckpt_filepath}", map_location=device)
+        if torch.cuda.device_count() > 1:
+            unet.module.load_state_dict(checkpoint_unet["unet_state_dict"], strict=True)
+        else:
+            unet.load_state_dict(checkpoint_unet["unet_state_dict"], strict=True)
+        print(f"pretrained checkpoint {existing_ckpt_filepath} loaded.")
+
+    noise_scheduler = define_instance(args, "noise_scheduler")
+
+    check_data = first(train_loader)
+    z = check_data["image"].to(device)
+    scale_factor = 1 / torch.std(z)
+    print(f"Scaling factor set to {scale_factor}.")
+
+    dist.barrier()
+    dist.all_reduce(scale_factor, op=torch.distributed.ReduceOp.AVG)
+    print(f"Rank {local_rank}: scale_factor -> {scale_factor}.")
+
+    optimizer = torch.optim.Adam(params=unet.parameters(), lr=lr)
+    print(f"optimizer -> Adam; lr -> {lr}.")
+
+    total_steps = (num_epochs * len(train_loader.dataset)) / num_images_per_batch
+    print(f"total number of training steps: {total_steps}.")
+
+    lr_scheduler = torch.optim.lr_scheduler.PolynomialLR(optimizer, total_iters=total_steps, power=2.0)
+
+    loss_pt = torch.nn.L1Loss()
+    best_loss = 1e4
```

Comment:
This variable seems unused.

Reply:
Hi @mingxin-zheng, I have updated the notebook. Please proceed. Thanks.